src/Containers/no_duplicate_dict.jl

```diff
@@ -0,0 +1,40 @@
+#  Copyright 2017, Iain Dunning, Joey Huchette, Miles Lubin, and contributors
+#  This Source Code Form is subject to the terms of the Mozilla Public
+#  License, v. 2.0. If a copy of the MPL was not distributed with this
+#  file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+    struct NoDuplicateDict{K, V} <: AbstractDict{K, V}
+        dict::Dict{K, V}
+    end
+
+Same as `Dict{K, V}` but errors if constructed from an iterator with duplicate
+keys.
+"""
+struct NoDuplicateDict{K, V} <: AbstractDict{K, V}
```

Comment:
Do we need to define a new type that's only used in one function? What about a function that takes a generator, constructs a `Dict` with errors on duplicates, and returns `Dict`?

Reply:
> constructs a `Dict` with errors on duplicates

How do you do that ? That's exactly `NoDuplicateDict` is trying to do

Reply:
I didn't dig too deeply, but my question was why you can't define a function like:
```julia
function dict_no_duplicates{K, V}(it) where {K, V}
    dict = Dict{K, V}()
    for (k, v) in it
        if haskey(dict, k)
            error(...)
        else
            dict[k] = v
        end
    end
    return dict
end
```

This doesn't require defining a new type. Maybe I'm missing something related to `dict_with_eltype`.

Reply:
The issue is that we don't know the type of the keys and values. The challenge is that Julia wants to infer it in a way that's friendly with Julia compiler and inference system.
So it starts iterating, assumes the type is always the same so it build the container with the type of the first element and then if a new element does not fit, it creates a new container with a wider type, copy and continue.
This is what is done in `map` (that DenseAxisArray rely on) and `dict_with_eltype` (that SparseAxisArray rely on).